nucliadb_node/nucliadb_node/indexer.py

```diff
@@ -0,0 +1,337 @@
+    def _priority_id(self) -> tuple[int, int]:
+        # Priority based on message source and smaller seqid
+        source_priority = self.priorities[self.index_message.source]
+        return (source_priority, self.seqid)
+
+    def __lt__(self, other):
+        if not isinstance(other, WorkUnit):
+            return NotImplemented
+        return self._priority_id.__lt__(other._priority_id)
+
+
+class ConcurrentShardIndexer:
+    def __init__(self, writer: Writer):
+        self.writer = writer
+        self.indexers: dict[str, tuple[PriorityIndexer, asyncio.Task]] = {}
+
+    async def initialize(self):
+        self.storage = await get_storage(service_name=SERVICE_NAME)
+
+    async def finalize(self):
+        for _, task in self.indexers.items():
+            try:
+                await task
+            except Exception as exc:
+                logger.exception(
+                    "Indexer task raised an error while finalizing it", exc_info=exc
+                )
+        self.indexers.clear()
+
+    async def index_message_soon(self, msg: Msg):
+        work = WorkUnit.from_msg(msg)
+        await work.mpu.start()
+
+        logger.info(
+            f"Index message for shard {work.shard_id} is being enqueued",
+            extra={
+                "shard": work.shard_id,
+                "subject": msg.subject,
+                "reply": msg.reply,
+                "seqid": work.seqid,
+                "storage_key": work.index_message.storage_key,
+            },
+        )
+
+        await self.push_work_to_indexer(work)
+        CONCURRENT_INDEXERS_COUNT.set(
+            len(self.indexers), labels=dict(node=work.node_id)
+        )
+
+    async def push_work_to_indexer(self, work: WorkUnit):
+        exists = work.shard_id in self.indexers
+        if exists:
+            indexer, _ = self.indexers[work.shard_id]
+            await indexer.index_soon(work)
+        else:
+            indexer = PriorityIndexer(writer=self.writer, storage=self.storage)
+            await indexer.index_soon(work)
+            task = asyncio.create_task(indexer.work_until_finish())
+            task.add_done_callback(
+                partial(
+                    ConcurrentShardIndexer._indexer_done_callback, self, work.shard_id
+                )
+            )
+            self.indexers[work.shard_id] = (indexer, task)
+
+    def _indexer_done_callback(self, shard_id: str, task: asyncio.Task):
+        # autoremove indexer when finished
+        self.indexers.pop(shard_id)
+
+        # propagate errors
+        if task.exception() is not None:
+            raise task.exception()  # type: ignore
+
+
+class PriorityIndexer:
+    def __init__(self, writer: Writer, storage: Storage):
+        self.writer = writer
+        self.storage = storage
+        self.work_queue: asyncio.PriorityQueue = asyncio.PriorityQueue()
+        self.working = False
+
+    async def index_soon(self, work: WorkUnit):
+        await self.work_queue.put(work)
```

Comment:
maybe make this use `put_nowait` and make the function not async.

We do not have a max size set on the queue so there isn't any reason to use async version here

Reply:
Changed to sync